### PR TITLE
Added dependent elements

### PR DESCRIPTION
**Pull request by chives**
This feature allows highlighting menu items for nested/dependent admin elements with proper linking and parent access.

TODO:
- [x] generic dependent element classes with specs
- [x] Doctrine dependent element classes with specs
- [x] Behat features with dependent elements
- [ ] <strike>rethink template headers (maybe another PR?)</strike>

### CODE REVIEW

**Comment by norzechowicz** [comment]
I don't see why this PR is made against 2.0. If someting is deprecated since 2.1 then merging this into 2.0 make no sense. 
2.0 should be frozen for any changes, only fixes and patches, that can be merged as a 2.0.x. 
This PR should be made against master, later when you decide that master is stable enought you can release 2.1 (with improvements like this one) and that should finally end up with 3.0 release also from master

**Comment by chives** [comment]
This feature is obviously targeted at 2.1 release, so i believe 2.1 branch should be created to allow parallel 2.x and 3.0 development

**Comment by norzechowicz** [comment]
it should be created but not from 2.0, 2.0 should be frozen.

**Comment by JarekW** [comment]
do we need all these additional classes (`DependentFormElement`, `DependentListElement`) in hierarchy?  Interface and trait should do the job. Naming also  should be unified (parent, dependent).

**Comment by chives** [comment]
Using trait is not as elegant as extending class and it's harder to understand especially for beginners. Besides there is also one method (`getRouteParameters()`) that need to be extended regardless of using the trait. Omitting this would lead to strange behavior like wrong redirection after successful action. Extension of this method cannot be simply enforced by language constructs.

"Dependent" means it depends on something especially "parent" element in our case, so these two terms does not refer to the same item but are opposite sides of the same concept: parent <-> dependent relation.

File: spec/FSi/Bundle/AdminBundle/Doctrine/Admin/DependentListElementSpec.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        $currentRequest->get(DependentElement::REQUEST_PARENT_PARAMETER)->willReturn('parent_object_id');
+        $parentDataIndexer->getData('parent_object_id')->willReturn('parent_object');
+
+        $this->setRequestStack($requestStack);
+        $this->setParentElement($parentElement);
+
+        $this->getParentObject()->shouldReturn('parent_object');
+    }
+
+    function its_route_parameters_contain_parent_object_id_if_its_available(
```
Comment (szymach) [inline spec/FSi/Bundle/AdminBundle/Doctrine/Admin/DependentListElementSpec.php]: `contains`

**Comment by chives** [comment]
@fsi-open

File: Admin/CRUD/DependentBatchElement.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+abstract class DependentBatchElement extends GenericBatchElement implements DependentElement
+{
+    use DependentElementImpl;
+
+    public function getRouteParameters()
+    {
+        return array_merge(
+            parent::getRouteParameters(),
+            [DependentElement::REQUEST_PARENT_PARAMETER => $this->getParentObjectId()]
```
Comment (szymach) [inline Admin/CRUD/DependentBatchElement.php]: Maybe `PARENT_REQUEST_PARAMETER`?

File: Behat/Context/AdminContext.php [hunk 3]
```diff
@@ -9,6 +9,7 @@
 
 namespace FSi\Bundle\AdminBundle\Behat\Context;
 
+use Behat\Behat\Tester\Exception\PendingException;
```
Comment (szymach) [inline Behat/Context/AdminContext.php]: Unused

File: Behat/Context/DataContext.php [hunk 4]
```diff
@@ -152,6 +159,19 @@ public function followingNewsExistInDatabase(TableNode $table)
             if (isset($newsNode['Date']) && $newsNode['Date']) {
                 $news->setDate(\DateTime::createFromFormat('Y-m-d', $newsNode['Date']));
             }
+            if (isset($newsNode['Category']) && $newsNode['Category']) {
+                /** @var Category|null $category */
+                $category = $categoryRepository->findOneBy(['title' => $newsNode['Category']]);
+
+                if ($category === null) {
+                    throw new InvalidArgumentException(sprintf(
+                        'can\'t find category by title "%s"',
```
Comment (szymach) [inline Behat/Context/DataContext.php]: `Can't`

File: Behat/Page/CategoryNewsEdit.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            throw new \Exception('Unable to find page header');
+        }
+
+        return $this->getElement('page header')->getText();
+    }
+
+    protected function verifyPage()
+    {
+        if (!$this->has('css', '#page-header:contains("Edit element")')) {
+            throw new UnexpectedPageException(sprintf("%s page is missing \"New element\" header", $this->path));
```
Comment (szymach) [inline Behat/Page/CategoryNewsEdit.php]: `Edit element`

File: CHANGELOG-2.1.md [hunk 6]
```diff
@@ -2,6 +2,15 @@
 
 This document describes all the significant changes made between 1.1/2.0 and 2.1.
 
+## Added bunch of new "dependent" admin element classes
```
Comment (szymach) [inline CHANGELOG-2.1.md]: `Introduced dependent admin elements`

File: CHANGELOG-2.1.md [hunk 7]
```diff
@@ -2,6 +2,15 @@
...
+
+These new abstract classes allow easy creation of admin elements which are dependent
```
Comment (szymach) [inline CHANGELOG-2.1.md]: We have introduced a new mechanism for creating a hierarchy for elements. Now you can make one element dependent from another, which means that:

 - actions witch dependent element properly highglight parent element's menu item
 - dependent element's method have access to parent admin element and the parent object instance,
   which can be used in datasource / datagrid / form building or view template

All currently existing admin elements have received a dependent version of themselves for your use. Also, you can easily create your own dependent elements by simply implementing the [DependentElement](Admin/DependentElement.php)

File: Resources/doc/admin_dependent_elements.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  your dependent admin element will not be linked from it directly
+- Use `getParentObject()` method to access object instance for which dependent admin
+  element was invoked. It should be helpful to constrain datasource results to objects
+  associated with the parent object (i.e. only news objects assigned to category),
+  and/or construct new instance of child object during form creation (i.e. create news
+  object already assigned to specific category). The parent object can also be used in
+  dependent element's view template i.e. to display its name and/or some url in
+  breadcrumbs.
+
+## 4. CRUD is not the only thing you usually need
```
Comment (szymach) [inline Resources/doc/admin_dependent_elements.md]: `usually` is redundant.

File: spec/FSi/Bundle/AdminBundle/Doctrine/Admin/DependentListElementSpec.php (hunk 1 above)
Comment (chives) [inline spec/FSi/Bundle/AdminBundle/Doctrine/Admin/DependentListElementSpec.php]: parameter**s** contain

File: features/fixtures/project/src/FSi/FixturesBundle/Entity/Category.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ * @ORM\Table(name="category")
+ */
+class Category
+{
+    /**
+     * @ORM\Column(type="integer")
+     * @ORM\Id
+     * @ORM\GeneratedValue(strategy="AUTO")
+     */
+    protected $id;
```
Comment (chives) [inline features/fixtures/project/src/FSi/FixturesBundle/Entity/Category.php]: it's just a convention in all entities of test app

**Pull request merged**